plugins/inputs/proxmox/proxmox.go

```diff
@@ -0,0 +1,251 @@
+package proxmox
+
+import (
+	"encoding/json"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+)
+
+var sampleConfig = `
+  ## API connection configuration. The API token was introduced in Proxmox v6.2. Required permissions for user and token: PVEAuditor role on /.
+  base_url = "https://localhost:8006/api2/json"
+  api_token = "USER@REALM!TOKENID=UUID"
+
+  ## Optional TLS Config
+  # tls_ca = "/etc/telegraf/ca.pem"
+  # tls_cert = "/etc/telegraf/cert.pem"
+  # tls_key = "/etc/telegraf/key.pem"
+  ## Use TLS but skip chain & host verification
+  insecure_skip_verify = false
+
+  # HTTP response timeout (default: 5s)
+  response_timeout = "5s"
+`
+
+func (px *Proxmox) SampleConfig() string {
+	return sampleConfig
+}
+
+func (px *Proxmox) Description() string {
+	return "Provides metrics from Proxmox nodes (Proxmox Virtual Environment > 6.2)."
+}
+
+func (px *Proxmox) Gather(acc telegraf.Accumulator) error {
+	err := getNodeSearchDomain(px)
+	if err != nil {
+		return err
+	}
+
+	gatherLxcData(px, acc)
+	gatherQemuData(px, acc)
+
+	return nil
+}
+
+func (px *Proxmox) Init() error {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+	px.hostname = hostname
+
+	tlsCfg, err := px.ClientConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+	px.httpClient = &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsCfg,
+		},
+		Timeout: px.ResponseTimeout.Duration,
+	}
+
+	return nil
+}
+
+func init() {
+	px := Proxmox{
+		requestFunction: performRequest,
+	}
+
+	inputs.Add("proxmox", func() telegraf.Input { return &px })
+}
+
+func getNodeSearchDomain(px *Proxmox) error {
+	apiUrl := "/nodes/" + px.hostname + "/dns"
+	jsonData, err := px.requestFunction(px, apiUrl, http.MethodGet, nil)
+
+	var nodeDns NodeDns
+	err = json.Unmarshal(jsonData, &nodeDns)
+	if err != nil {
+		return err
+	}
+	px.nodeSearchDomain = nodeDns.Data.Searchdomain
+
+	return nil
+}
+
+func performRequest(px *Proxmox, apiUrl string, method string, data url.Values) ([]byte, error) {
+	request, err := http.NewRequest(method, px.BaseURL+apiUrl, strings.NewReader(data.Encode()))
+	if err != nil {
+		return nil, err
+	}
+	request.Header.Add("Authorization", "PVEAPIToken="+px.APIToken)
+
+	resp, err := px.httpClient.Do(request)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	responseBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return responseBody, nil
+}
+
+func gatherLxcData(px *Proxmox, acc telegraf.Accumulator) {
+	gatherVmData(px, acc, LXC)
+}
+
+func gatherQemuData(px *Proxmox, acc telegraf.Accumulator) {
+	gatherVmData(px, acc, QEMU)
+}
+
+func gatherVmData(px *Proxmox, acc telegraf.Accumulator, rt ResourceType) {
+	vmStats, err := getVmStats(px, rt)
+	if err != nil {
+		log.Fatal(err)
```

Comment:
I missed this the first time around. Don't use Fatal as it won't handle errors gracefully. Instead, add the following to the struct:
```
Log telegraf.Logger `toml:"-"`
```

At runtime it'll be given a logger instance, and you can use it to do `px.Log.Errorf("Error getting vm stats: %v", err)`

You can optionally update the tests to set to set the logger instance with a test logger: `px.Log = testutil.Logger{}` so they don't error when a log line is called.

Reply:
After this I think you're good.